### PR TITLE
Fixes #11963: Please, legitimate "Priority" usage in multi-instance directives

### PR DESCRIPTION
**Pull request by fanf**
https://www.rudder-project.org/redmine/issues/11963

### CODE REVIEW

File: 4_advanced_usage/35_directive_ordering.txt [hunk 1]
```diff
@@ -2,16 +2,76 @@
...
-When policies are generated for each nodes, all Directives based on a same Technique are merged together, preventing a complete ordering of all Directives. However, Rudder uses a best-effort method for ordering Directives, based on alphanumeric ordering.
+In Rudder, there is two separated ways to order things, depending about what is the resulting policy. So, before that, we need to explain how Policies are generated on the
+agent. 
 
-Rudder tries to order first on Rules name, then on Directives name.
+==== Policy generation and Directive merge
 
-Example:
+In Rudder, Policies are generated from Directives, but several Directives based on the same Technique always lead to *one* Policy on the agent. 
+For unique (non multi-instance) Technique, the most propritary is selected. For multi-instance Technique, the different Directive values are *merged*
```
Comment (ncharles) [inline 4_advanced_usage/35_directive_ordering.txt]: typo: "the one with the high priority is selected"

File: 4_advanced_usage/35_directive_ordering.txt [hunk 2]
```diff
@@ -2,16 +2,76 @@
 
 Configuration in Rudder are based on desired states, describing the expected state of the system. However, there are cases where having order is desirable (like ensuring that a JVM is present before deploying an Application server, or ensuring a user is present before setting it sudoers), even if it will converge over the course of several agent runs.
 
-When policies are generated for each nodes, all Directives based on a same Technique are merged together, preventing a complete ordering of all Directives. However, Rudder uses a best-effort method for ordering Directives, based on alphanumeric ordering.
+In Rudder, there is two separated ways to order things, depending about what is the resulting policy. So, before that, we need to explain how Policies are generated on the
```
Comment (ncharles) [inline 4_advanced_usage/35_directive_ordering.txt]: "depending on the type of Technique" ?

File: 4_advanced_usage/35_directive_ordering.txt [hunk 3]
```diff
@@ -2,16 +2,76 @@
...
+ See https://www.rudder-project.org/redmine/issues/10625[Don't merge directive from same technique on generation] for more information.
+=====
+
+
+==== Sorting Directives based on the *same* Technique
+
+For Directive based on the same Technique, the sort order is based on the *Priority* value of the Directive. Between two Directive, the one with the highest *Priority*
+is the first:
+
+- for a *non* multi-instance Technique, it means that it is the one that is chosen as the Policy,
```
Comment (ncharles) [inline 4_advanced_usage/35_directive_ordering.txt]: it means that it is there is only one that is chosen in the resulting Policies (the others are discared),

File: 4_advanced_usage/35_directive_ordering.txt [hunk 4]
```diff
@@ -2,16 +2,76 @@
...
+=====
+
+
+==== Sorting Directives based on the *same* Technique
+
+For Directive based on the same Technique, the sort order is based on the *Priority* value of the Directive. Between two Directive, the one with the highest *Priority*
+is the first:
+
+- for a *non* multi-instance Technique, it means that it is the one that is chosen as the Policy, 
+- for a multi-instance Technique, it means that its variable set will be declared and check first in the corresponding Policy.
```
Comment (ncharles) [inline 4_advanced_usage/35_directive_ordering.txt]: I'm not sure of the actually meaning there ...

File: 4_advanced_usage/35_directive_ordering.txt [hunk 5]
```diff
@@ -2,16 +2,76 @@
...
+
+==== Sorting Directives based on the *same* Technique
+
+For Directive based on the same Technique, the sort order is based on the *Priority* value of the Directive. Between two Directive, the one with the highest *Priority*
+is the first:
+
+- for a *non* multi-instance Technique, it means that it is the one that is chosen as the Policy, 
+- for a multi-instance Technique, it means that its variable set will be declared and check first in the corresponding Policy. 
+
+For equals *Prioriy* for several *Directive*, the Rule Name, and then the Directive name is used in alphanumeric order.
```
Comment (ncharles) [inline 4_advanced_usage/35_directive_ordering.txt]: If several *Directives* have the same *Priority*, the Rule name, and then the Directive name are used for storing in alphanumeric order
(note: the priority has precedence over Rule/Directive name ordrering ? I thought it was first Rule/Directive name to choose the Techniques order, then Priority to sort within them, for non unique Technique)

File: 4_advanced_usage/35_directive_ordering.txt [hunk 6]
```diff
@@ -2,16 +2,76 @@
...
+
+===== Special use case: overriding generic_variable_definition
+
+You can use the merging of Directive to define variable override with the "Generic Variable Definition" Technique. 
+
+For example, let say you want to define a *DNS* variable with default value *[default dns]* and on some node case, 
+a value *[overrided dns]*:
+
+- Create a Directive [1] with *high* priority: it will be your *default* case, so set *DNS* to *[default dns]*.
+- Create an other Directive [2] with *lower* priority: it will be you specialized case, so set *DNS* to *[overrided dns]*.
```
Comment (ncharles) [inline 4_advanced_usage/35_directive_ordering.txt]: *lower* as in "1", or has in *10* ?

File: 4_advanced_usage/35_directive_ordering.txt [hunk 7]
```diff
@@ -2,16 +2,76 @@
...
+
+Then, a node with only Directive [1] will have the default value defined, and a node with both Directives will have the overrided one.
```
Comment (ncharles) [inline 4_advanced_usage/35_directive_ordering.txt]: typo: the overriding one

File: 4_advanced_usage/35_directive_ordering.txt [hunk 8]
```diff
@@ -2,16 +2,76 @@
...
+
+Rudder uses a best-effort method for ordering Policies, based on alphanumeric ordering of the corresponding Rule, then Directive name. 
+
+When several Directive were merged, Rudder choose the first (Rule name, Directive name) as the ordering value to use for the resulting Policy. 
+
+
+.Best practice
+[TIP]
+=====
+You should always start Rules and Directives name by 2 (or 3) digits to be able to easely reorder Policy evaluation if the need happen:
```
Comment (ncharles) [inline 4_advanced_usage/35_directive_ordering.txt]: typo: easily

File: 4_advanced_usage/35_directive_ordering.txt (hunk 5 above)
Comment (fanf) [reply 4_advanced_usage/35_directive_ordering.txt]: only when merging / chosing the directive for an unique technique

**Comment by fanf** [comment]
OK, merging this PR

**Pull request merged**